experiments/esp32_mp/src/main.c

```diff
@@ -0,0 +1,131 @@
+#include <zephyr.h>
+#include <misc/printk.h>
+
+#define _REG(base, off) (*(volatile u32_t *)((base) + (off)))
+
+#define RTC_CNTL_BASE             0x3ff48000
+#define RTC_CNTL_OPTIONS0     _REG(RTC_CNTL_BASE, 0x0)
+#define RTC_CNTL_SW_CPU_STALL _REG(RTC_CNTL_BASE, 0xac)
+
+#define DPORT_BASE                 0x3ff00000
+#define DPORT_APPCPU_CTRL_A    _REG(DPORT_BASE, 0x02C)
+#define DPORT_APPCPU_CTRL_B    _REG(DPORT_BASE, 0x030)
+#define DPORT_APPCPU_CTRL_C    _REG(DPORT_BASE, 0x034)
+
+/* Two fields with same naming conventions live in two different
+ * registers and have different widths...
+ */
+#define RTC_CNTL_SW_STALL_APPCPU_C0  (0x03 << 0)  /* reg: OPTIONS0 */
+#define RTC_CNTL_SW_STALL_APPCPU_C1  (0x3f << 20) /* reg: SW_CPU_STALL */
+
+#define DPORT_APPCPU_CLKGATE_EN  BIT(0)
+#define DPORT_APPCPU_RUNSTALL    BIT(0)
+#define DPORT_APPCPU_RESETTING   BIT(0)
+
+
+/* These calls are ROM-resident and have fixed addresses. No, I don't
+ * know how they plan on updating these portably either.
+ */
+typedef void (*esp32rom_call_t)(int);
+static const esp32rom_call_t esp32rom_Cache_Flush = (void*)0x40009a14;
+static const esp32rom_call_t esp32rom_Cache_Read_Enable = (void*)0x40009a84;
+static const esp32rom_call_t esp32rom_ets_set_appcpu_boot_addr = (void*)0x4000689c;
+
+volatile int counter;
+volatile int ccount1;
+volatile void *stack1;
+
+/* Carefully constructed to use no stack beyond compiler-generated ABI
+ * instructions.  WE DO NOT KNOW WHERE THE STACK FOR THIS FUNCTION IS.
+ * The ROM library just picks a spot on its own with no input from our
+ * app linkage and tells us nothing about it until we're already
+ * running.
+ *
+ * This is also a problem for stack switching: we don't know what the
+ * register window state might be either.  Remember Xtensa registers
+ * are lazy-filled: it's not impossible (though in practice probably
+ * not likely) that access to a4-a15 in the swapped stack will cause
+ * an exception that tries to "fill" them from a stack that no longer
+ * exists.  Need to research how to properly reset the register window
+ * (which is different from existing context switch code, which can
+ * presume a working stack on which to spill!).
+ */
+void appcpu_entry(void)
+{
+	__asm__ volatile("s32i sp, %0, 0" : "=a"(stack1));
+	counter = 1000;
+	while (1) {
+		static volatile int dummy;
+		for (dummy = 0; dummy < 1000000; dummy++);
+		__asm__ volatile("rsr.ccount %0" : "=a"(ccount1));
+		counter++;
+	}
+}
+
+/* The calls and sequencing here were extracted from the ESP-32
```

Comment:
Have you taken a look at how nuttx does this? Is it much different from FreeRTOS?

Reply:
It's basically identical, yeah.